File: DataStructures/7-Tree/Tree.cpp
#include <iostream>
#include <vector>
#include <queue>

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class TreeAlgorithms {
public:
    // DFS: In-order Traversal (Left, Root, Right)
    void inorderTraversal(TreeNode* root) {
        if (root == nullptr) return;
        inorderTraversal(root->left);
        std::cout << root->val << " ";
        inorderTraversal(root->right);
    }

    // DFS: Pre-order Traversal (Root, Left, Right)
    void preorderTraversal(TreeNode* root) {
        if (root == nullptr) return;
        std::cout << root->val << " ";
        preorderTraversal(root->left);
        preorderTraversal(root->right);
    }

    // DFS: Post-order Traversal (Left, Right, Root)
    void postorderTraversal(TreeNode* root) {
        if (root == nullptr) return;
        postorderTraversal(root->left);
        postorderTraversal(root->right);
        std::cout << root->val << " ";
    }

    // BFS: Level-order Traversal
    void levelorderTraversal(TreeNode* root) {
        if (root == nullptr) return;
        std::queue<TreeNode*> q;
        q.push(root);
        while (!q.empty()) {
            TreeNode* current = q.front();
            q.pop();
            std::cout << current->val << " ";
            if (current->left != nullptr) {
                q.push(current->left);
            }
            if (current->right != nullptr) {
                q.push(current->right);
            }
        }
    }
};

// Example Usage
int main() {
    /*
     * Example Tree:
     *      4
     *     / \
     *    2   7
     *   / \
     *  1   3
     */
    TreeNode* root = new TreeNode(4);
    root->left = new TreeNode(2);
    root->right = new TreeNode(7);
    root->left->left = new TreeNode(1);
    root->left->right = new TreeNode(3);

    TreeAlgorithms ta;

    std::cout << "In-order Traversal (DFS): ";
    ta.inorderTraversal(root);
    std::cout << std::endl; // Expected: 1 2 3 4 7

    std::cout << "Pre-order Traversal (DFS): ";
    ta.preorderTraversal(root);
    std::cout << std::endl; // Expected: 4 2 1 3 7

    std::cout << "Post-order Traversal (DFS): ";
    ta.postorderTraversal(root);
    std::cout << std::endl; // Expected: 1 3 2 7 4

    std::cout << "Level-order Traversal (BFS): ";
    ta.levelorderTraversal(root);
    std::cout << std::endl; // Expected: 4 2 7 1 3

    // Clean up memory (important in C++)
    delete root->left->left;
    delete root->left->right;
    delete root->left;
    delete root->right;
    delete root;

    return 0;
}

File: DataStructures/7-Tree/Tree.swift
import Foundation

// Definition for a binary tree node.
public class TreeNode {
    public var val: Int
    public var left: TreeNode?
    public var right: TreeNode?
    public init(_ val: Int) {
        self.val = val
        self.left = nil
        self.right = nil
    }
}

class TreeAlgorithms {
    // DFS: In-order Traversal (Left, Root, Right)
    func inorderTraversal(_ root: TreeNode?) {
        guard let root = root else { return }
        inorderTraversal(root.left)
        print(root.val, terminator: " ")
        inorderTraversal(root.right)
    }

    // DFS: Pre-order Traversal (Root, Left, Right)
    func preorderTraversal(_ root: TreeNode?) {
        guard let root = root else { return }
        print(root.val, terminator: " ")
        preorderTraversal(root.left)
        preorderTraversal(root.right)
    }

    // DFS: Post-order Traversal (Left, Right, Root)
    func postorderTraversal(_ root: TreeNode?) {
        guard let root = root else { return }
        postorderTraversal(root.left)
        postorderTraversal(root.right)
        print(root.val, terminator: " ")
    }

    // BFS: Level-order Traversal
    func levelorderTraversal(_ root: TreeNode?) {
        guard let root = root else { return }
        var queue: [TreeNode] = [root]
        
        while !queue.isEmpty {
            let current = queue.removeFirst()
            print(current.val, terminator: " ")
            
            if let leftChild = current.left {
                queue.append(leftChild)
            }
            if let rightChild = current.right {
                queue.append(rightChild)
            }
        }
    }
}

// Example Usage
/*
 * Example Tree:
 *      4
 *     / \
 *    2   7
 *   / \
 *  1   3
 */
let root = TreeNode(4)
root.left = TreeNode(2)
root.right = TreeNode(7)
root.left?.left = TreeNode(1)
root.left?.right = TreeNode(3)

let ta = TreeAlgorithms()

print("In-order Traversal (DFS): ", terminator: "")
ta.inorderTraversal(root)
print() // Expected: 1 2 3 4 7

print("Pre-order Traversal (DFS): ", terminator: "")
ta.preorderTraversal(root)
print() // Expected: 4 2 1 3 7

print("Post-order Traversal (DFS): ", terminator: "")
ta.postorderTraversal(root)
print() // Expected: 1 3 2 7 4

print("Level-order Traversal (BFS): ", terminator: "")
ta.levelorderTraversal(root)
print() // Expected: 4 2 7 1 3
